`RealmDictionary` feature

edualonso opened the pull request:
Closes #537

### Dictionary API
A `RealmDictionary` is a `Map<K, V>`, where keys are always strings and values are Realm primitives or objects, be it regular or embedded. Dictionaries can also be expressed as:
- `MutableSet<MutableMap.MutableEntry<K, V>>` by calling `RealmDictionary.entries`. The entry set is a live collection connected to its underlying dictionary. Deletions and insertions are allowed and will result in an updated dictionary. Individual entries can also be modified and will result in an update key-value dictionary pair. Unmanaged entries can also be instantiated via factory - see "Factories" section below. Managed entries are also live key-value pairs and mutating them will update the dictionary accordingly.
- `MutableSet<K>` by calling `RealmDictionary.keys`. The key set is a live collection connected to its underlying dictionary. Deletions are allowed though insertions aren't - it isn't possible to insert a key without a value. Core delivers the keys as `realm_results_t` but subscribing for key changes is not exposed for now.
- `MutableCollection<V>` by calling `RealmDictionary.values`. The value collection is a live collection connected to its underlying dictionary. Deletions are allowed though insertions aren't - it isn't possible to insert a value without a key. Core delivers the values as `realm_results_t` so queries on dictionary objects are funnelled through these.

### `RealmDictionary` and `RealmMap` abstractions
Thanks to how the compiler plugin extracts information from the type system it is (almost) possible to work at a `RealmMap` level, though we aren't exposing a public implementation of it and `RealmMap` fields aren't allowed either.

To make the implementation more readable and less painful, a number of type aliases have been added:
- `internal typealias RealmMapEntrySet<K, V> = MutableSet<MutableMap.MutableEntry<K, V>>`
- `internal typealias RealmMapMutableEntry<K, V> = MutableMap.MutableEntry<K, V>`
- `public typealias RealmDictionaryEntrySet<V> = RealmMapEntrySet<String, V>`
- `public typealias RealmDictionaryMutableEntry<V> = RealmMapMutableEntry<String, V>`

### Object dictionaries
Object dictionaries may contain either regular or embedded objects and `null` values are allowed. `RealmDictionary` of objects must be nullable - not the field itself, which is not allowed for any type of Realm collection, but the collection type instead.

### Factories
Unmanaged dictionaries can be instantiated by calling the `realmDictionaryOf` functions that allow the following input values:
- `realmDictionaryOf(vararg elements: Pair<String, T>)`
- `realmDictionaryOf(elements: Collection<Pair<String, T>>)`

Unmanaged `RealmDictionaryMutableEntry` instances can be used to add key-value pairs to the dictionary through the dictionary's entry set and can be created by calling the `realmDictionaryEntryOf` functions which accept the following input values:
- `realmDictionaryEntryOf(pair: Pair<String, V>)`
- `realmDictionaryEntryOf(key: String, value: V)`
- `realmDictionaryEntryOf(entry: Map.Entry<String, V>)`

### Queries on dictionaries
There is no direct way to parse a query on a dictionary object through the C-API. However, we can use the `realm_results_t` object yielded by the C-API when calling `RealmDictionary.values` to parse a query on the results.

### Notifications and change sets
A dictionary change set contains:
- the deleted keys
- the inserted keys
- the changed keys

Pull request plan:
- [x] basic infrastructure and API: `RealmMap` and `RealmDictionary` interfaces and associated helper functions (stubs)
- [x] setting up exhaustive testing infrastructure, add support for `copyToRealm` for objects with dictionaries and interface functions `get`, `put`, `putAll`, `remove` and `clear` for all types except `RealmAny` (it will be done once the interface implementation is complete)
- [x] support for `entries` (i.e. `RealmMapEntrySet`, `UnmanagedRealmMapEntry` and `ManagedRealmMapEntry`) and internal entry set iterator (which enables iterating through the dictionary's k-v pairs externally)
- [x] `containsKey`, `containsValue`~~- currently missing support in the C-API (https://github.com/realm/realm-core/issues/6181)~~
- [x] `values` - this change entails refactoring `RealmResults` to contain primitive values
- [x] support for `RealmAny`
- [ ] support for ~~`Deleteable` and (dictionaries don't own objects like sets and lists do so removing them from the dictionary won't remove them from the realm)~~ `asFlow` - ~~currently blocked by https://github.com/realm/realm-core/issues/6228~~
- [x] allow `ByteArray`s to be compared structurally to allow `remove` operations on entry set and values
- [x] support for `copyFromRealm`
- [x] support for dynamic realms
- [x] support for backlinks
- [x] sync testing
- [x] aggregation queries on dictionaries and add missing test cases for sets and lists
- [x] embedded objects - https://github.com/realm/realm-kotlin/pull/1269
- [x] queries on dictionaries - https://github.com/realm/realm-kotlin/pull/1269
- [x] notifications - https://github.com/realm/realm-kotlin/pull/1274
- [x] documentation
- [x] changelog entry

clementetb commented on packages/test-base/src/commonMain/kotlin/io/realm/kotlin/test/util/TypeDescriptor.kt:
```diff
@@ -335,7 +345,8 @@ object TypeDescriptor {
-    // TODO Dict
+    val allDictionaryFieldTypes = elementTypesForDictionary.filter { it.realmFieldType.dictionarySupport }
+        .map { RealmFieldType(CollectionType.RLM_COLLECTION_TYPE_DICTIONARY, it) }
     val allFieldTypes: List<RealmFieldType> = allSingularFieldTypes + allListFieldTypes
```
allFieldTypes is missing allSetFieldTypes and allDictionaryFieldTypes

edualonso commented:
Still missing the changelog entry and some documentation but otherwise ready for final review.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/query/RealmElementQuery.kt:
```diff
@@ -17,8 +17,13 @@
 package io.realm.kotlin.query
 
 import io.realm.kotlin.Deleteable
+import io.realm.kotlin.MutableRealm
```
All these new imports don't seem to be used?

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmMap.kt:
```diff
@@ -0,0 +1,67 @@
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * TODO
```
Missing documentation. We can probably copy most of it from Java

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmMap.kt:
```diff
@@ -0,0 +1,67 @@
+public interface RealmMap<K, V> : MutableMap<K, V>
+
+/**
+ * TODO
```
Missing documentation

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmMap.kt:
```diff
@@ -0,0 +1,67 @@
+/**
+ * TODO
+ */
+public interface RealmDictionary<E> : RealmMap<String, E> {
```
Nitpick, but should we use `<String, V>` instead of `<String, E>`? So it matches the RealmMap names.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmMap.kt:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return a flow representing changes to the dictionary.
+     */
+    public fun asFlow(): Flow<DictionaryChange<E>>
```
Would it make sense to add ` public fun asFlow(): Flow<MapChange<K, V>>` to RealmMap already? And then this needs to be `override public fun asFlow(): Flow<DictionaryChange<E>>`

edualonso replied:
After some more reflecting I think it's better to just have this function in `RealmMap` and not even override it for `RealmDictionary`. The reason for this is that the aliases I added for the notification changes are pretty useless 😞 The IDE doesn't take them into account when telling it to explicitly state the type and uses the parent type instead. So it might seem like a good idea to use aliases for when the types are a mouthful but it seems they useless when trying to get the IDE to use them for convenience. So I'm more inclined to delete them now.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmMap.kt:
```diff
@@ -0,0 +1,67 @@
+     * the elements in a timely manner the coroutine scope will be cancelled with a
+     * [CancellationException].
+     *
+     * @return a flow representing changes to the dictionary.
```
Missing throws statements.

cmelchior reviewed (commented):
Halfway through review, minor comments only so far.

cmelchior reviewed (commented):
Only missing 3000 lines of tests now. Looking good so far 😄

cmelchior commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:
```diff
@@ -0,0 +1,2966 @@
+
+    @Test
+    fun unmanaged() {
+        // No need to be exhaustive here, just checking delegation works
```
Why don't we need to be exhaustive? We should be sure we know how unmanaged objects work as well. That said, It would not block merging this PR for it. Maybe just create an issue and we can add testing for it later.

cmelchior commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:
```diff
@@ -0,0 +1,2966 @@
+    val realm: Realm
+
+    override fun toString(): String
+    fun copyToRealm()
```
Looks like we are missing testing for `equals/hashCode` for the `RealmDictionary` itself?

edualonso replied:
True, I will add them 👍

edualonso commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/QueryTests.kt:
```diff
@@ -2019,9 +2023,206 @@ class QueryTests {
+        listOf("@sum", "@min", "@max", "@avg")
+            .forEach { aggregator ->
+                listOf(
+                    QuerySample::intListField.name, // Just test one Long value
```
I have created https://github.com/realm/realm-kotlin/issues/1295

edualonso commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/BacklinksTests.kt:
```diff
@@ -278,24 +336,29 @@ class BacklinksTests {
             val child = this.copyToRealm(Child())
 
-            val parents = Array(5) {
+            val parents = (1..5).map {
```
The problem with `repeat` is that it doesn't do any mapping and we would have to add the output manually for each iteration. Otherwise I agree, it would look better.

cmelchior reviewed (commented):
Done...I think we are missing some corner cases around equality testing for Dictionary but otherwise it looks good 🚀

edualonso commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:
```diff
@@ -0,0 +1,2966 @@
+    }
+
+    @Test
+    fun closedRealm_readFails() {
```
I think I simply forgot about them 🙈 

These will also most likely start to fail once we remove the call to `RealmReference.checkClosed()` so I'll add a TODO.

edualonso commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:
```diff
@@ -0,0 +1,2966 @@
+    }
+
+    @Test
+    fun entries_equals() {
```
Do you mean comparing two `RealmDictionary.entries.toString()` or two individual `entry.toString()`?

cmelchior replied:
It looks like we are missing all of them.

edualonso commented on packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:
```diff
@@ -0,0 +1,2966 @@
+    }
+
+    @Test
+    fun values_equals() {
```
Same as with the comment about the entry set. Do you mean the value collection as a whole or individual values?

edualonso commented:
Added missing `toString` tests and added scaffolding for future `equals` and `hashCode` tests for when https://github.com/realm/realm-kotlin/issues/1097 is fixed

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:
```diff
@@ -657,6 +677,11 @@ internal class ManagedRealmDictionary<V> constructor(
         }
     }
+
+    override fun toString(): String = entries.joinToString { (key, value) -> "[$key,$value]" }
```
Hmm, I can see where this is coming from, but it looks pretty dangerous, i.e. lets say you have dictionaries with 100's or 1000's of elements. I would probably avoid doing this, but there doesn't seem to be a great alternative either.

Since `toString` is generally used for debugging I would probably rather turn this into something like `RealmDictionary{owner=<classType>,objKey=<objectId>,version=<transactionVersion>}`. It would be easy to compute and you would still get some context for debugging.

edualonso replied:
Since there was no `toString` implementation for the other collections I took the approach we use for Java lists but I agree it can easily backfire with large datasets. I think your proposal is fair, I would only add `size` for clarity too.

cmelchior reviewed (changes requested):
Deferring tests to we fix the `equals` issue is fine, but I think the `toString()` implementation is making too many assumptions about dictionaries not having many elements. So we should change it to something else IMO. I proposed a solution in the comment, but happy to discuss alternatives.

rorbech reviewed (approved):
Finally had a look on this. Haven't gone through all the details but only the overall structures and essential mechanisms ... And looks good to me.
